feat: use new generics format (#1133)
GetStream/stream-chat-react-native: docusaurus/docs/reactnative/customization/typescript.mdx

It is important that the proper generics be applied to the `stream-chat-js` client when it is instantiated.
The [documentation on `stream-chat-js` TypeScript](https://github.com/GetStream/stream-chat-js#typescript-v2xx) has examples of how this can be done in detail.
The client takes seven generics that correspond to the seven customizable fields that currently exist in `stream-chat-js`.

```tsx
const client = StreamChat.getInstance<StreamChatType>('YOUR_API_KEY');
```

where `StreamChatType` can be defined as a type with the seven generics that correspond to the seven customizable fields as follows:

```tsx
type StreamChatType = {
  attachmentType: LocalAttachmentType;
  channelType: LocalChannelType;
  commandType: LocalCommandType;
  eventType: LocalEventType;
  messageType: LocalMessageType;
  reactionType: LocalReactionType;
  userType: LocalUserType;
};
```

:::note

> `LocalAttachmentType`, `LocalChannelType` etc. are type deifinitions for their respective key as per your application types necessities.

Author: ```suggestion
`LocalAttachmentType`, `LocalChannelType` etc. are type definitions for their respective key as per your application types necessities.
```

};
```

> Additional fields on the defaults i.e. `image` are custom fields used by `stream-chat-react-native` already within the SDK.

Author: ```suggestion
Additional fields on the defaults i.e. `image` is a custom fields used by `stream-chat-react-native` already within the SDK.
```

Core to understanding this usage is [how generics can be applied to JSX elements](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-9.html#generic-type-arguments-in-jsx-elements).

In many cases the use of a single prop such as `client` or `channel` allows TypeScript to infer the generics on an element.

> In this case `StreamChatType` is inferred from `channel` and passed to the props type for a custom Attachment component.

Author: Indeed you are correct

> In this case `StreamChatType` is inferred from `channel` and passed to the props type for a custom Attachment component.

Author: ```suggestion
In this case `channelType` within `StreamChatType` is inferred from `channel` and passed to the props type for a custom Attachment component.
```
